src/lib/nested_property.js

```diff
@@ -113,24 +115,36 @@ function npGet(cont, parts) {
 }
 
 /*
- * Check known non-data-array arrays (containers). Data arrays only contain scalars,
- * so parts[end] values, such as -1 or n, indicate we are not dealing with a dataArray.
- * The ONLY case we are looking for is where the entire array is selected, parts[end] === 'x'
- * AND the replacement value is an array.
+ * Can this value be deleted? We can delete any empty object (null, undefined, [], {})
+ * EXCEPT empty data arrays. Info arrays can be safely deleted, but not deleting them
+ * has no ill effects other than leaving a trace or layout object with some cruft in it.
+ * But deleting data arrays can change the meaning of the object, as `[]` means there is
+ * data for this attribute, it's just empty right now while `undefined` means the data
+ * should be filled in with defaults to match other data arrays. So we do some simple
+ * tests here to find known non-data arrays but don't worry too much about not deleting
+ * some arrays that would actually be safe to delete.
  */
-function isDataArray(val, key) {
-
-    var containers = ['annotations', 'shapes', 'range', 'domain', 'buttons'],
-        isNotAContainer = containers.indexOf(key) === -1;
-
-    return isArray(val) && isNotAContainer;
+var INFO_PATTERNS = /(^|.)((domain|range)(\.[xy])?|args|parallels)$/;
```

Comment:
added even more complexity https://github.com/plotly/plotly.js/pull/1403/commits/c6378b12e0d2cd99e7fc2541a6dfcf3b4597fdc2 after discussion with @rreusser but opened #1410 to discuss simplification later on.